RFC: Add an interfaces manual chapter

mbauman opened the pull request:
Document the iterable, indexable, and abstract array interfaces.

I'm open to suggestions on the tables.  I like the way they summarize the requirements, and I think the iterable and indexable ones work well, but the array table is nearly unreadable since it needs to scroll horizontally.

mbauman committed:
Add an interfaces manual chapter

Document the iterable, indexable, and abstract array interfaces.

[av skip]

mbauman commented:
There's room at the bottom for someone else to later add information about the Equality interface (`==`/`isequal`/`hash`/`<`/`isless`) and Display interface (`display`/`show`/`print`).  :)

tkelman commented:
This is great! Should the code blocks be doctests here?

mbauman commented:
Ah, yes, thank you, they should.

tkelman commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,223 @@
+ Interfaces
+************
+
+A lot of the power and extensibility in Julia comes from a collection of informal interfaces.  By extending few specific methods to work for a custom type, objects of that type not only receive those functionalities, but they are also able to be used in other methods that are written to generically build upon those behaviors.
```
By extending a few`

IainNZ commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,223 @@
+:func:`next(iter, state) <next>`                           Returns the current item and the next state
+:func:`done(iter, state) <done>`                           Tests if there are any items remaining
+**Important optional methods**    **Default definition**   **Brief description**
+:func:`eltype(IterType) <eltype>` ``Any``                  The container's element type
```
I wouldn't use the word container like this, or at least I don't think of all iterables as being containers. Maybe the "The element type of the container or sequence`. Its a really minor point though

mbauman replied:
How about "the type of each item returned by `next`"?  Describing it as the element type is a bit circular.

IainNZ commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,223 @@
+:func:`length(iter) <length>`     (*undefined*)            The container's length
+================================= ======================== ===========================================
+
+Sequential iteration is implemented by the methods :func:`start`, :func:`done`, and :func:`next`. Instead of mutating objects as they are iterated over, Julia provides these three methods to keep track of the iteration state externally from the object. The :func:`start(iter)` method returns an initial ``state`` object that gets passed along to :func:`done(iter, state)`, which tests if there are a ...
```
returns the initial ``state`` for the iterable object ``iter`` that

IainNZ commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,223 @@
+      # body
+    end
+
+is translated into::
```
I think `lowered` is the technical word, but translated probably more familiar to more people

IainNZ commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,223 @@
+Any object that has these three methods appropriately defined can be used in a ``for`` loop since the syntax::
+
+    for i in iter   # or  "for i = iter"
+      # body
```
Four space indent?

IainNZ commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,223 @@
+    julia> mean(Squares(100)), std(Squares(100))
+    (3383.5,3024.355854282583)
+
+There are a few more methods we can extend to give Julia more information about this iterable collection.  We know that the elements in a ``Squares`` collection will always be ``Int``. By extending the :func:`eltype` method, we can give that information to Julia and help it make more specialized code in the more complicated methods. We also know the number of elements in our collection, so we can ...
```
"``Squares`` sequence"?

mbauman replied:
Ah, yes, I like sequence much better in general.

IainNZ commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,223 @@
+
+While we can rely upon generic implementations, we can also extend specific methods where we know there is a simpler algorithm.  For example, there's a formula to compute the sum of squares, so we can override the generic iterative version with a more performant solution::
+
+    julia> sum(S::Squares) = (n = S.count; return n*(n+1)*(2n+1)÷6)
```
Does this need to be Base.sum?

mbauman replied:
Keen eye.  It sure does.  `make doctest` found that one, too.

IainNZ replied:
Hah, I was reading closely, I've been waiting for this document for a long time. Now we just need one for the show/print zoo...

IainNZ commented:
Conceptually, what separates these from a "formal" interface? That the type isn't directly annotated with something that says "Iterable"?

IainNZ commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,223 @@
+:func:`getindex(X, i) <getindex>`      ``X[i]``, indexed element access
+:func:`setindex!(X, v, i) <setindex!>` ``X[i] = v``, indexed assignment
+:func:`endof(X) <endof>`               The last index, used in ``X[end]``
+====================================== ==================================
```
What happens with colons, like `Squares[:]`? Should that go here?

mbauman replied:
Yeah, as I was writing it I briefly considered using that as a motivating segue into the AbstractArray section. Not sure why I didn't end up doing that… I guess I got distracted. `S[:]` now gets lowered to `getindex(S, Colon())`.  So it's entirely up to the type to define the appropriate method.  Definitely worth a mention.

mbauman committed:
Add doctests

mbauman commented:
> Conceptually, what separates these from a "formal" interface?  That the type isn't directly annotated with something that says "Iterable"?

Yes, I think so.  See https://github.com/JuliaLang/julia/issues/6975.  I guess it gets a little blurry for the `AbstractArray` case, since there behaviors are much more defined by it being an `AbstractArray`.  That is, iterables are duck-typed.  AbstractArrays have their own method signature.

mbauman committed:
Update after review

* Use sequence instead of collection in some places
* Change eltype and length descriptions
* Indentation fixes
* Talk about how defining getindex manually is a game of whackamole without support from AbstractArray
* Link to Array Indexing section

[av skip]

johnmyleswhite commented:
This is really awesome. Well done, @mbauman.

nalimilan commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,278 @@
+:func:`similar(A, ::Type{S}, dims::NTuple{Int}) <similar>` ``Array(S, dims)``                           Return a mutable array with the specified element type and dimensions
+========================================================== ============================================ =======================================================================================
+
+If a type is defined as a subtype of ``AbstractArray``, it inherits a very large set of complicated behaviors including iteration and multidimensional indexing built on top of single-element access.
```
"Complicated" doesn't make it very appealing. :-) Maybe you could make it more obvious that you *benefit* from "rich" automatic fallbacks by simply defining basic methods?

nalimilan commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,278 @@
+
+If a type is defined as a subtype of ``AbstractArray``, it inherits a very large set of complicated behaviors including iteration and multidimensional indexing built on top of single-element access.
+
+A key part in defining an ``AbstractArray`` subtype is :func:`Base.linearindexing`. Since indexing is such an important part of an array and often occurs in hot loops, it's important to make both indexing and indexed assignment as efficient as possible.  Array data structures are typically defined in one of two ways: either it's most efficient to access the elements using just one index (using li ...
```
In "it intrinsically accesses", the "it" sounds a bit weird, as it parallels the previous "it's" which does not refer to anything in particular. Maybe use the passive "or elements are intrinsically accessed"?

mbauman replied:
Good catch.  This is a bad habit of mine.  See my update — I thought it read nicer to change the first half to agree with the second.

nalimilan commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,278 @@
+           Base.linearindexing(::Type{SquaresVector}) = Base.LinearFast()
+           Base.getindex(S::SquaresVector, i::Int) = i*i;
+
+Note that it's very important to specify the two parameters of the ``AbstractArray``; the first defines the :func:`eltype`, and the second defines the :func:`ndims`.  But that's it takes for our squares type to be an iterable, indexable, and completely functional array:
```
"that's *what* it takes"?

nalimilan commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,278 @@
+           Base.setindex!{T}(A::SparseArray{T,2}, v, i1::Int, i2::Int)          = (A.data[(i1,i2)] = v)
+           Base.setindex!{T}(A::SparseArray{T,3}, v, i1::Int, i2::Int, i3::Int) = (A.data[(i1,i2,i3)] = v);
+
+Notice that this is a ``LinearSlow`` array, so we must manually define :func:`getindex` and :func:`setindex!` for each dimensionality we'd like to support.  Unlike the ``SquaresVector``, we are able to define :func:`setindex!`, and so we can mutate the array:
```
Maybe you should way a word about this difference between ``LinearSlow`` and ``LinearFast`` right after the table above?

mbauman replied:
Yes, more can definitely be said.

nalimilan commented on doc/manual/interfaces.rst:
```diff
@@ -0,0 +1,278 @@
+     1.0  4.0  7.0
+     2.0  5.0  8.0
+
+And now, in addition to all the iterable and indexable methods from above, these types can interact with eachother and use all the methods defined in the standard library for ``AbstractArrays``:
```
"each other"

nalimilan commented:
Really cool! Don't you think the titles of the sections could be changed to `Iterable`, `Indexable`, and `AbstractArray`s?

mbauman committed:
Update after Milan's comments

* complicated -> rich
* fix linearindexing description coherence between phrases by changing the first part of the phrase
* add a paragraph about how linearindexing impacts which getindex method(s) must get defined
* Fix missing *what*
* each other

[av skip]

mbauman committed:
More cross-links!

[av skip]

nalimilan commented:
Thanks, looks good!

mbauman commented:
Thanks @nalimilan!  Updated (oh, you already found it as I was writing this!).  I like the words iteration and indexing much better than iterable and indexable, but if others think the latter vocabulary is clearer I'll happily change it.

I've also added more cross-links to and from other sections of the manual and standard library.  In doing so, I ran across http://docs.julialang.org/en/latest/manual/arrays/#implementation.  That page and section are written more in a descriptive how-to-use-it fashion instead of the way this is more of a how-to-implement-it.  I think they can both live together nicely, but perhaps some of the information could be consolidated.  Thoughts?

Also, I'm very tempted to document `unsafe_getindex`.

And I really cannot wait for #11242 to get merged.  A syntax is really needed to talk about these things simply and clearly.

mbauman commented:
I'll merge this now.  It's documentation and there are clamors for hashing/equality, comparison, and display interfaces.  I don't have time to add those myself in the short term, but maybe once there's an obvious place for them to live others will be able to add it.

mbauman commented:
Shoot, sorry, should have squashed.

ScottPJones commented:
There is no requirement that an AbstractArray implement a hash function, is there?
The arrays I've implemented in the past frequently didn't use hashing at all (they used a B+tree or a patricia trie).

mbauman commented:
They get the default implementation for free.  Should be added.
